### PR TITLE
catalog: Use durable trait for catalog open

### PR DESCRIPTION
**Pull request by jkosh44**
This commit updates the catalog open/initialization process to
use the new durable catalog trait.

Works towards resolving #20953

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -5058,47 +5061,71 @@ impl Catalog {
...
+                tls,
+            },
+            now,
+        )
+        .await
     }
 
-    /// Opens a debug catalog from a stash.
-    pub async fn open_debug_stash(stash: Stash, now: NowFn) -> Result<Catalog, anyhow::Error> {
+    pub async fn open_debug_stash_catalog_config(
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:5104]: The names of these debug methods aren't great, so I'm open to suggestions.

**Review by def- — COMMENTED** [review]
I assume this is supposed to have no externally visible effect, thus no new tests required. Nightly already ran, I triggered coverage too: https://buildkite.com/materialize/coverage/builds/228

File: src/catalog/src/stash.rs [hunk 2]
```diff
@@ -50,148 +52,295 @@ use crate::transaction::{
...
+    }
 
-            let mut conn = Connection { stash };
+    async fn get_deployment_generation(&mut self) -> Result<Option<u64>, Error> {
+        let stash = match &mut self.stash {
+            None => match self.open_stash_read_only().await {
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(None),
+                Err(e) => return Err(e.into()),
+            },
```
Comment (def-) [inline src/catalog/src/stash.rs:193]: The code looks relatively harmless, but this appears uncovered

File: src/catalog/src/stash.rs [hunk 3]
```diff
@@ -50,148 +52,295 @@ use crate::transaction::{
...
+    loop {
+        match open_inner(stash, now.clone(), bootstrap_args, deploy_generation).await {
+            Ok(conn) => {
+                return Ok(conn);
+            }
+            Err((given_stash, err)) => {
+                stash = given_stash;
+                let should_retry = matches!(&err, Error::Stash(se) if se.should_retry());
+                if !should_retry || retry.next().await.is_none() {
+                    return Err(err);
```
Comment (def-) [inline src/catalog/src/stash.rs:237]: Similar argument, might be interesting to have a test that exercises this error case to see that the error is correctly propagated all the way.

**Review by def- — APPROVED** [review]
Coverage looks mostly fine, marked two spots, not sure if they are easy to test.

File: src/catalog/src/stash.rs (hunk 2 above)
Comment (jkosh44) [reply src/catalog/src/stash.rs:197]: I added some tests that should cover this.

File: src/catalog/src/stash.rs (hunk 3 above)
Comment (jkosh44) [reply src/catalog/src/stash.rs:241]: This is hard to exercise because we have to mock a CRDB issue which is not easy.

File: src/catalog/src/lib.rs [hunk 4]
```diff
@@ -176,28 +178,56 @@ pub struct BootstrapArgs {
...
+    /// but never durably commits them. This is used to check and see if
+    /// opening the catalog would be successful, without making any durable
+    /// changes.
     ///
     /// Will return an error in the following scenarios:
-    ///   - Catalog not initialized.
+    ///   - Catalog initialization fails.
     ///   - Catalog migrations fail.
-    async fn check_open(&self) -> Result<(), Error>;
+    async fn open_check(
```
Comment (ParkMyCar) [inline src/catalog/src/lib.rs:192]: 🚲🚲🚲 bikeshed alert 🚲🚲🚲

When I saw "check" I thought read-only, which this API sort of is, what do you think about just `open_savepoint`? Totally optional and not part of this PR so really up to you

File: src/catalog/src/stash.rs [hunk 5]
```diff
@@ -50,148 +52,299 @@ use crate::transaction::{
...
-            conn
+#[tracing::instrument(name = "storage::open_inner", level = "info", skip_all)]
+async fn open_inner(
+    mut stash: Stash,
+    now: NowFn,
+    bootstrap_args: &BootstrapArgs,
+    deploy_generation: Option<u64>,
+) -> Result<Connection, (Stash, Error)> {
+    // Initialize the Stash if it hasn't been already
+    let mut conn = if !match stash.is_initialized().await {
```
Comment (ParkMyCar) [inline src/catalog/src/stash.rs:256]: small thing, it might be nice to re-write this using `let-else` to avoid the "right drift", e.g.

```
let Ok(is_init) = stash.is_initialized().await else {
  return Err((stash, e.into());
};
```

You can probably also do this for all of the match statements in this function. It's just code formatting though so absolutely no need to do it if you don't want to!

**Review by ParkMyCar — APPROVED** [review]

File: src/catalog/src/lib.rs (hunk 4 above)
Comment (jkosh44) [reply src/catalog/src/lib.rs]: I'm not in love with "check" either. I wanted to avoid "savepoint" to avoid tying it to the CRDB implementation, but I think it is probably better than check. For now I've switched back to "savepoint" unless we can come up with something better.
